contributors/design-proposals/sig-cli/apply_v2.md

```diff
@@ -0,0 +1,201 @@
+2. Only set / delete fields defined in the file either currently or in during last update from the file
+  - Retain updates to fields made by controllers that were never managed in the file
+3. Merge fields based on the field name, merge lists based on a field contained within the list elements
+
+The apply logic has been difficult to maintain from a contributor perspective, and difficult
+to understand from a user perspective.  Much of this pain has comes from the `apply` command out
+growing its implementation.  In order to continue to support apply and provide better user visibility
+into what it is doing, the command must be re-architected to break client dependencies on the server-side
+implementation, address gaps in diff-patch semantics preventing correctly applying certain fields,
+and to organize the code in a modular and extendable manner.
+
+Reducing the code complexity and improving the correctness will also make it simpler to move this logic
+to the server as part of an API in the future, while also addressing the immediate needs of the client.
+
+## Changes
+
+Implementation of proposed changes under review in PR [52349](https://github.com/kubernetes/kubernetes/pull/52349)
+
+### Use read-update instead of patch
+
+#### Why
+
+Issues with updating the object by calculating and sending a patch to the server
+
+- Patch API is not fully featured enough to represent all of the operations required
+  to update an object and frequently has to be expanded with new side-directives describing
+  how the server should apply the patch.  Updates are done in the form of adding new
+  merge directives to the patch format on the server, and invoking them from the client.
+- Having to update both client and server to fix bugs introduces challenges to supporting
+  version skew, as the client will depend on the server supporting a patch semantic.  There
+  is currently no versioning of the patch semantics, so the client cannot reliably know what
+  patch semantics the server supports.
+- Patch syntax is not human readable, making it hard for humans to evaluate the end state
+  that will result from the patch.  Implementing a diff operation has been exceedingly
+  complex.
+  - Humans must mentally apply directives (delete, retain keys, ordering) in order to understand
+    the result of the patch.
+  - Merge keys are implicit (not shown in the patch), but change the result
+  - Merge strategy is implicit (not show in the patch), but change the result
+  - Merge keys and strategy are incorrect and buggy for several object fields (e.g. Container.Ports)
+    resulting in counter-intuitive results (and there is no diff command to help users see this
+    beforehand)
+  - Patch hides values on the remote object, which may have hidden interactions with the patch values
+    (e.g. Deployment.Spec.RolloutStrategy)
+- Generating a patch to make a change is more complicated than simply making the change to
```

Comment:
It seems to me that the client vs. server discussion isn't the main point. 

The main point is whether the merge is accomplished via separate diff and patch passes, or with a custom merge implementation.

Reply:
And the new implementation should be mostly reusable for a server-side apply API.

Reply:
Yes exactly.  I can change the name of the doc if that would help.